geoutils/georaster/raster.py

```diff
@@ -195,31 +210,97 @@ def __init__(
-            # Save the absolute on-disk filename
-            self.filename = os.path.abspath(filename_or_dataset)
-            if as_memfile:
-                # open the file in memory
-                memfile = MemoryFile(open(filename_or_dataset, "rb"))
-                # Read the file as a rasterio dataset
-                self.ds = memfile.open()
+        elif isinstance(filename_or_dataset, (str, rio.io.DatasetReader, rio.io.MemoryFile)):
+
+            with ExitStack():
+                if isinstance(filename_or_dataset, str):
+                    ds: rio.io.DatasetReader = rio.open(filename_or_dataset)
+                    self.filename = filename_or_dataset
+                elif isinstance(filename_or_dataset, rio.io.DatasetReader):
+                    ds = filename_or_dataset
+                    self.filename = filename_or_dataset.files[0]
+                else:
+                    ds = filename_or_dataset.open()
+                    self.filename = None
+
+                self.transform = ds.transform
+                self.crs = ds.crs
+                self.nodata = ds.nodata
+                self.name = ds.name
+                self.driver = ds.driver
+                self.tags.update(ds.tags())
+
+                self._disk_shape = (ds.count, ds.height, ds.width)
+                self._disk_indexes = ds.indexes
+                self._disk_dtypes = ds.dtypes
+
+                if attrs is not None:
+                    for attr in attrs:
+                        self.__setattr__(attr, ds.__getattr__(attr))
+
+            # Check number of bands to be loaded
+            if bands is None:
+                nbands = self.nbands
+            elif isinstance(bands, int):
+                nbands = 1
+            elif isinstance(bands, abc.Iterable):
+                nbands = len(bands)
+
+            # Downsampled image size
+            if not isinstance(downsample, (int, float)):
+                raise ValueError("downsample must be of type int or float")
+            if downsample == 1:
+                out_shape = (nbands, self.height, self.width)
             else:
-                self.ds = rio.open(filename_or_dataset, "r")
-
-        # If rio.Dataset is passed
-        elif isinstance(filename_or_dataset, rio.io.DatasetReader):
-            self.filename = filename_or_dataset.files[0]
-            self.ds = filename_or_dataset
-
-        # Or, image is already a Memory File.
-        elif isinstance(filename_or_dataset, rio.io.MemoryFile):
-            self.ds = filename_or_dataset.open()
+                down_width = int(np.ceil(self.width / downsample))
+                down_height = int(np.ceil(self.height / downsample))
+                out_shape = (nbands, down_height, down_width)
+                res = tuple(np.asarray(self.res) * downsample)
+                self.transform = rio.transform.from_origin(self.bounds.left, self.bounds.top, res[0], res[1])
+
+            if load_data:
```

Comment:
there seem to be two identical if statement in a row!

Reply:
It's just to make sure it's actually actually loaded.





Just kidding, I'll fix it!

Reply:
Fixed in b059bfb